Source/WebDriver/soup/WebSocketServerSoup.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebSocketServer.h"
+
+#include "CommandResult.h"
+#include <cstdio>
+#include <libsoup/soup-websocket-connection.h>
+#include <libsoup/soup.h>
+#include <optional>
+#include <tuple>
+#include <wtf/glib/GUniquePtr.h>
+#include <wtf/text/WTFString.h>
+
+namespace WebDriver {
+
+static bool soupServerListen(SoupServer* server, const String& host, unsigned port, GError** error)
+{
+    static const auto options = static_cast<SoupServerListenOptions>(0);
+
+    if (!host || host == "local"_s)
+        return soup_server_listen_local(server, port, options, error);
+
+    if (host == "all"_s)
+        return soup_server_listen_all(server, port, options, error);
+
+    GRefPtr<GSocketAddress> address = adoptGRef(g_inet_socket_address_new_from_string(host.utf8().data(), port));
+    if (!address) {
+        g_set_error(error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT, "Invalid host IP address '%s'", host.utf8().data());
+        return false;
+    }
+
+    return soup_server_listen(server, address.get(), options, error);
+}
+
+static void handleIncomingHandshake(SoupServer* server, SoupServerMessage* message, const char* path, GHashTable*, gpointer userData)
+{
+    if (!soup_websocket_server_check_handshake(message, nullptr, nullptr, nullptr, nullptr)) {
+        soup_server_message_set_status(message, 500, "Invalid request");
+        return;
+    }
+    GRefPtr<SoupServerMessage> protectedMessage = message;
+    soup_server_pause_message(server, message);
```

Comment:
Do we really need to pause the message? This method is sync.

Reply:
Removed.